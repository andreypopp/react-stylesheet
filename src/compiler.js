/**
 * @flow
 */

import hyphenateStyleName from 'fbjs/lib/hyphenateStyleName';
import memoizeStringOnly from 'fbjs/lib/memoizeStringOnly';
import computeHash from 'murmurhash-js/murmurhash3_gc';

import type {StylesheetSpec} from './index';
import PseudoClassSet from './PseudoClassSet';
import UnitlessNumberPropSet from './UnitlessNumberPropSet';

export type CompiledStylesheet = {
  id: string,
  rules: Array<{selector: string, props: string}>,
  spec: StylesheetSpec,
};

export function compile(spec: StylesheetSpec): CompiledStylesheet {
  const displayName = spec.displayName || 'Component';
  const rules = [];
  let id = [];
  for (const key in spec) {
    if (key === 'displayName') {
      continue;
    }
    // compile variant
    const variant = [];
    compileVariant(variant, [], spec[key]);
    // now turn variant into rule
    const hash = computeHash(displayName + '||' + JSON.stringify(variant));
    id.push(hash);
    for (let i = 0; i < variant.length; i++) {
      const rule = variant[i];
      if (rule.props.length === 0) {
        continue;
      }
      // selector
      let selector = displayName;
      if (key !== 'base') {
        selector = `${selector}-${key}`;
      }
      selector = `${selector}-${hash}`;
      if (rule.rightToLeft) {
        selector = `${selector}.RTL`;
      }
      if (rule.selector.length > 0) {
        selector = `${selector}:${rule.selector.join(':')}`;
      }
      // props
      const props = rule.props.join(';\n') + ';';
      rules.push({selector, props});
    }
  }

  return {
    id: computeHash(id.join('||')),
    rules,
    spec,
  };
}

function compileVariant(rules, selector, variant) {
  const props = [];
  const propsRTL = [];
  const variantRules = [];
  for (const key in variant) {
    const val = variant[key];
    if (PseudoClassSet[key]) {
      compileVariant(variantRules, selector.concat(key), val);
    } else {
      compileProperty(props, propsRTL, key, val);
    }
  }
  if (props.length > 0) {
    rules.push({selector, props, rightToLeft: false});
  }
  if (propsRTL.length > 0) {
    rules.push({selector, props: propsRTL, rightToLeft: true});
  }
  if (variantRules.length > 0) {
    rules.push(...variantRules);
  }
}

function compileProperty(propsLTR, propsRTL, name, value) {
  switch (name) {
    case 'paddingStart': {
      const cvalue = compileValue('padding', value);
      propsLTR.push(`padding-left: ${cvalue}`);
      propsRTL.push(`padding-right: ${cvalue}`);
      break;
    }
    case 'paddingEnd': {
      const cvalue = compileValue('padding', value);
      propsLTR.push(`padding-right: ${cvalue}`);
      propsRTL.push(`padding-left: ${cvalue}`);
      break;
    }
    case 'paddingVertical': {
      const cvalue = compileValue('padding', value);
      propsLTR.push(`padding-top: ${cvalue}`);
      propsLTR.push(`padding-bottom: ${cvalue}`);
      break;
    }
    case 'paddingHorizontal': {
      const cvalue = compileValue('padding', value);
      propsLTR.push(`padding-left: ${cvalue}`);
      propsLTR.push(`padding-right: ${cvalue}`);
      break;
    }
    case 'marginStart': {
      const cvalue = compileValue('margin', value);
      propsLTR.push(`margin-left: ${cvalue}`);
      propsRTL.push(`margin-right: ${cvalue}`);
      break;
    }
    case 'marginEnd': {
      const cvalue = compileValue('margin', value);
      propsLTR.push(`margin-right: ${cvalue}`);
      propsRTL.push(`margin-left: ${cvalue}`);
      break;
    }
    case 'marginVertical': {
      const cvalue = compileValue('margin', value);
      propsLTR.push(`margin-top: ${cvalue}`);
      propsLTR.push(`margin-bottom: ${cvalue}`);
      break;
    }
    case 'marginHorizontal': {
      const cvalue = compileValue('margin', value);
      propsLTR.push(`margin-left: ${cvalue}`);
      propsLTR.push(`margin-right: ${cvalue}`);
      break;
    }
    case 'startOffset': {
      const cvalue = compileValue('left', value);
      propsLTR.push(`left: ${cvalue}`);
      propsRTL.push(`right: ${cvalue}`);
      break;
    }
    case 'endOffset': {
      const cvalue = compileValue('right', value);
      propsLTR.push(`right: ${cvalue}`);
      propsRTL.push(`left: ${cvalue}`);
      break;
    }
    default:
      const cname = compileName(name);
      const cvalue = compileValue(name, value);
      propsLTR.push(`${cname}: ${cvalue}`);
  }
}

export function compileValue(name: string, value: mixed): string {
  let isNonNumeric = isNaN(value);
  if (
    isNonNumeric ||
    value === 0 ||
    (UnitlessNumberPropSet.hasOwnProperty(name) && UnitlessNumberPropSet[name])
  ) {
    return '' + ((value: any): string); // cast to string
  } else {
    return ((value: any): string) + 'px';
  }
}

const isEmpty = value => value == null || value === '' || value === false;
export const compileName = memoizeStringOnly(name => hyphenateStyleName(name));
