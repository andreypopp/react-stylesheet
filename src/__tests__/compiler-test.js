/**
 * @flow
 */

import * as Compiler from '../compiler';

test('compileValue()', function() {
  expect(Compiler.compileValue('width', 1)).toBe('1px');
  expect(Compiler.compileValue('flex', 1)).toBe('1');
});

test('compileName()', function() {
  expect(Compiler.compileName('width')).toBe('width');
  expect(Compiler.compileName('flex')).toBe('flex');
  expect(Compiler.compileName('borderLeft')).toBe('border-left');
  expect(Compiler.compileName('border-left')).toBe('border-left');
});

test('compile() variants', function() {
  expect(
    Compiler.compile({
      base: {
        color: 'red',
      },
    }),
  ).toMatchSnapshot();
  expect(
    Compiler.compile({
      something: {
        color: 'red',
      },
    }),
  ).toMatchSnapshot();
});

test('compile() w/ pseudoclasses', function() {
  expect(
    Compiler.compile({
      base: {
        color: 'red',
        hover: {
          color: 'black',
        },
      },
    }),
  ).toMatchSnapshot();
  expect(
    Compiler.compile({
      base: {
        color: 'red',
        hover: {
          color: 'black',
          focus: {
            color: 'yellow',
          },
        },
      },
    }),
  ).toMatchSnapshot();
  expect(
    Compiler.compile({
      something: {
        color: 'red',
        hover: {
          color: 'black',
        },
      },
    }),
  ).toMatchSnapshot();
});

test('compile() paddingStart/paddingEnd', function() {
  expect(
    Compiler.compile({
      base: {
        paddingStart: 12,
      },
    }),
  ).toMatchSnapshot();
  expect(
    Compiler.compile({
      base: {
        paddingEnd: 12,
      },
    }),
  ).toMatchSnapshot();
});

test('compile() paddingVertical/paddingHorizontal', function() {
  expect(
    Compiler.compile({
      base: {
        paddingVertical: 12,
      },
    }),
  ).toMatchSnapshot();
  expect(
    Compiler.compile({
      base: {
        paddingHorizontal: 12,
      },
    }),
  ).toMatchSnapshot();
});

test('compile() marginStart/marginEnd', function() {
  expect(
    Compiler.compile({
      base: {
        marginStart: 12,
      },
    }),
  ).toMatchSnapshot();
  expect(
    Compiler.compile({
      base: {
        marginEnd: 12,
      },
    }),
  ).toMatchSnapshot();
});

test('compile() marginVertical/marginHorizontal', function() {
  expect(
    Compiler.compile({
      base: {
        marginVertical: 12,
      },
    }),
  ).toMatchSnapshot();
  expect(
    Compiler.compile({
      base: {
        marginHorizontal: 12,
      },
    }),
  ).toMatchSnapshot();
});

test('compile() startOffset/endOffset', function() {
  expect(
    Compiler.compile({
      base: {
        startOffset: 12,
      },
    }),
  ).toMatchSnapshot();
  expect(
    Compiler.compile({
      base: {
        endOffset: 12,
      },
    }),
  ).toMatchSnapshot();
});
