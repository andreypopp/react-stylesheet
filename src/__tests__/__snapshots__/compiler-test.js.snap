// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile() marginStart/marginEnd 1`] = `
Object {
  "id": 3854612195,
  "rules": Array [
    Object {
      "props": "margin-left: 12px;",
      "selector": "Component-4100871942",
    },
    Object {
      "props": "margin-right: 12px;",
      "selector": "Component-4100871942.RTL",
    },
  ],
  "spec": Object {
    "base": Object {
      "marginStart": 12,
    },
  },
  "variantToClassName": Object {
    "base": "Component-4100871942",
  },
}
`;

exports[`compile() marginStart/marginEnd 2`] = `
Object {
  "id": 3489670315,
  "rules": Array [
    Object {
      "props": "margin-right: 12px;",
      "selector": "Component-1522872987",
    },
    Object {
      "props": "margin-left: 12px;",
      "selector": "Component-1522872987.RTL",
    },
  ],
  "spec": Object {
    "base": Object {
      "marginEnd": 12,
    },
  },
  "variantToClassName": Object {
    "base": "Component-1522872987",
  },
}
`;

exports[`compile() marginVertical/marginHorizontal 1`] = `
Object {
  "id": 1807630354,
  "rules": Array [
    Object {
      "props": "margin-top: 12px;
margin-bottom: 12px;",
      "selector": "Component-2127533042",
    },
  ],
  "spec": Object {
    "base": Object {
      "marginVertical": 12,
    },
  },
  "variantToClassName": Object {
    "base": "Component-2127533042",
  },
}
`;

exports[`compile() marginVertical/marginHorizontal 2`] = `
Object {
  "id": 1388465887,
  "rules": Array [
    Object {
      "props": "margin-left: 12px;
margin-right: 12px;",
      "selector": "Component-2856744843",
    },
  ],
  "spec": Object {
    "base": Object {
      "marginHorizontal": 12,
    },
  },
  "variantToClassName": Object {
    "base": "Component-2856744843",
  },
}
`;

exports[`compile() paddingStart/paddingEnd 1`] = `
Object {
  "id": 2782882808,
  "rules": Array [
    Object {
      "props": "padding-left: 12px;",
      "selector": "Component-1749670398",
    },
    Object {
      "props": "padding-right: 12px;",
      "selector": "Component-1749670398.RTL",
    },
  ],
  "spec": Object {
    "base": Object {
      "paddingStart": 12,
    },
  },
  "variantToClassName": Object {
    "base": "Component-1749670398",
  },
}
`;

exports[`compile() paddingStart/paddingEnd 2`] = `
Object {
  "id": 419523180,
  "rules": Array [
    Object {
      "props": "padding-right: 12px;",
      "selector": "Component-1864011970",
    },
    Object {
      "props": "padding-left: 12px;",
      "selector": "Component-1864011970.RTL",
    },
  ],
  "spec": Object {
    "base": Object {
      "paddingEnd": 12,
    },
  },
  "variantToClassName": Object {
    "base": "Component-1864011970",
  },
}
`;

exports[`compile() paddingVertical/paddingHorizontal 1`] = `
Object {
  "id": 1772353193,
  "rules": Array [
    Object {
      "props": "padding-top: 12px;
padding-bottom: 12px;",
      "selector": "Component-3658477728",
    },
  ],
  "spec": Object {
    "base": Object {
      "paddingVertical": 12,
    },
  },
  "variantToClassName": Object {
    "base": "Component-3658477728",
  },
}
`;

exports[`compile() paddingVertical/paddingHorizontal 2`] = `
Object {
  "id": 3288951138,
  "rules": Array [
    Object {
      "props": "padding-left: 12px;
padding-right: 12px;",
      "selector": "Component-1749467061",
    },
  ],
  "spec": Object {
    "base": Object {
      "paddingHorizontal": 12,
    },
  },
  "variantToClassName": Object {
    "base": "Component-1749467061",
  },
}
`;

exports[`compile() startOffset/endOffset 1`] = `
Object {
  "id": 2257361630,
  "rules": Array [
    Object {
      "props": "left: 12px;",
      "selector": "Component-4111961331",
    },
    Object {
      "props": "right: 12px;",
      "selector": "Component-4111961331.RTL",
    },
  ],
  "spec": Object {
    "base": Object {
      "startOffset": 12,
    },
  },
  "variantToClassName": Object {
    "base": "Component-4111961331",
  },
}
`;

exports[`compile() startOffset/endOffset 2`] = `
Object {
  "id": 2960308138,
  "rules": Array [
    Object {
      "props": "right: 12px;",
      "selector": "Component-2672255561",
    },
    Object {
      "props": "left: 12px;",
      "selector": "Component-2672255561.RTL",
    },
  ],
  "spec": Object {
    "base": Object {
      "endOffset": 12,
    },
  },
  "variantToClassName": Object {
    "base": "Component-2672255561",
  },
}
`;

exports[`compile() variants 1`] = `
Object {
  "id": 2594978201,
  "rules": Array [
    Object {
      "props": "color: red;",
      "selector": "Component-2808132786",
    },
  ],
  "spec": Object {
    "base": Object {
      "color": "red",
    },
  },
  "variantToClassName": Object {
    "base": "Component-2808132786",
  },
}
`;

exports[`compile() variants 2`] = `
Object {
  "id": 2594978201,
  "rules": Array [
    Object {
      "props": "color: red;",
      "selector": "Component-something-2808132786",
    },
  ],
  "spec": Object {
    "something": Object {
      "color": "red",
    },
  },
  "variantToClassName": Object {
    "something": "Component-something-2808132786",
  },
}
`;

exports[`compile() w/ pseudoclasses 1`] = `
Object {
  "id": 865716088,
  "rules": Array [
    Object {
      "props": "color: red;",
      "selector": "Component-2281268226",
    },
    Object {
      "props": "color: black;",
      "selector": "Component-2281268226:hover",
    },
  ],
  "spec": Object {
    "base": Object {
      "color": "red",
      "hover": Object {
        "color": "black",
      },
    },
  },
  "variantToClassName": Object {
    "base": "Component-2281268226",
  },
}
`;

exports[`compile() w/ pseudoclasses 2`] = `
Object {
  "id": 164222277,
  "rules": Array [
    Object {
      "props": "color: red;",
      "selector": "Component-147224304",
    },
    Object {
      "props": "color: black;",
      "selector": "Component-147224304:hover",
    },
    Object {
      "props": "color: yellow;",
      "selector": "Component-147224304:hover:focus",
    },
  ],
  "spec": Object {
    "base": Object {
      "color": "red",
      "hover": Object {
        "color": "black",
        "focus": Object {
          "color": "yellow",
        },
      },
    },
  },
  "variantToClassName": Object {
    "base": "Component-147224304",
  },
}
`;

exports[`compile() w/ pseudoclasses 3`] = `
Object {
  "id": 865716088,
  "rules": Array [
    Object {
      "props": "color: red;",
      "selector": "Component-something-2281268226",
    },
    Object {
      "props": "color: black;",
      "selector": "Component-something-2281268226:hover",
    },
  ],
  "spec": Object {
    "something": Object {
      "color": "red",
      "hover": Object {
        "color": "black",
      },
    },
  },
  "variantToClassName": Object {
    "something": "Component-something-2281268226",
  },
}
`;
